### DISCUSSION TITLE
Pi4K V2

### DISCUSSION THREAD

**Discussion by mhashim6** [Pi4J/pi4j#239] [Ideas]
Hi everyone,
I'm posting this here as I'm starting on implementing a new Kotlin DSL for V2 of Pi4J.
Years ago I created [Pi4K](https://github.com/mhashim6/Pi4K) as a complement to Pi4J V1, and hopefully, we can make things better for Kotlin fans on V2.

I'll be posting about which interfaces and behaviors I'll be targetting (think contexts, builders, IO...etc). And hopefully, together we can make a quite inclusive, intuitive, and fun-to-use DSL.

Thank you!

**Comment by FDelporte** [comment]
Hi and welcome to this discussion forum :-)

I'm a total newbie on Kotlin apart from some very basic experiments. So what would it take me to create my first KotlinOnRaspberryPi project? Having such a description would be wonderful. In the past, I created such basic examples with Java, JBang, Spring/Vaadin,... all based on the same wiring as found in the description of a minimal example application with button and LED (https://pi4j.com/getting-started/minimal-example-application/). So from my point of view, this would be a perfect first step to being able and knowing how to do the same with Kotlin.

**Reply by mhashim6** [reply]
Thank you @FDelporte!
I've already started on the interface. I'm excited to share that there's enough progress on the DSL to make a beautiful minimal example. Here's a fully-functioning snippet that is identical in behaviour to the java version (I stripped the comments for brievity):

``` kotlin
import com.pi4j.io.gpio.digital.*
import com.pi4j.pi4k.digital.*
import com.pi4j.pi4k.*

private const val PIN_BUTTON = 24 // PIN 18 = BCM 24
private const val PIN_LED = 22 // PIN 15 = BCM 22
private var pressCount = 0

fun main() {
    console {
        title("<-- The Pi4J Project -->", "Minimal Example project")
        withinAutoContext {
            printLoadedPlatforms(this)
            printDefaultPlatform(this)
            printProviders(this)
            
            digitalInput(PIN_BUTTON) {
                id("button")
                name("Press button")
                pull(PullResistance.PULL_DOWN)
                debounce(3000L)
                piGpioProvider()
            }.listen {
                if (it.state() == DigitalState.LOW) {
                    pressCount++
                    +"Button was pressed for the ${pressCount}th time"
                }
            }

            digitalOutput(PIN_LED) {
                id("led")
                name("LED Flasher")
                shutdown(DigitalState.LOW)
                initial(DigitalState.LOW)
                piGpioProvider()
            }.run {
                printRegistry(this@withinAutoContext)
                while (pressCount < 5) {
                    if (equals(DigitalState.HIGH)) {
                        +"LED low"
                        low()
                    } else {
                        +"LED high"
                        high()
                    }
                    Thread.sleep((500 / (pressCount + 1)).toLong())
                }
            }
            shutdown()
        }
    }
}

```

There's more room for making this even more concise and pretty. Also, I created other DSL for more advnaced user cases that I'm happy to share once ready.

**Reply by mhashim6** [reply]
Changes I'm working on:
- `withinAutoContext` should be `pi4j`
- `Context::shutdown` can be implicitly called at the end of a `pi4j` block
- Explicit listeners on High & Low states
- `ContextBuilder` DSL
- `Platform` DSL

**Reply by DieterHolz** [reply]
That looks really promising.

Would you like to share the repository with the library functions?

**Comment by mhashim6** [comment]
@DieterHolz Sure!
I created a new [repo](https://github.com/Pi4J/pi4j-kotlin) for v2.
https://github.com/Pi4J/pi4j-kotlin
